doc/compiler/jitserver/Caching.md

```diff
@@ -0,0 +1,49 @@
+
+# Why caching is important
+When server is compiling a method, it needs to know a lot of information that is located on the client side, e.g. the addresses of RAM classes, GC mode, class hierarchy information, interpreter profiling data and much more. To obtain this information, server will make a remote call over the network to the client, client will find it, and send it back to the server. During the course of a compilation server will make hundreds of such calls (on average), assuming no caching is involved.
+Why is it bad? First of all, remote calls are expensive, both in terms of latency and CPU, especially in terms of CPU. If the server makes hundreds of requests to the client just for one compilation, client will spend more CPU time just sending and receiving data over the network than it would take for it to compile a method locally. What makes it worse is that in order for the client to fetch the requested data it needs to do some work, which additionally increases CPU consumption.
+Caching helps alleviate this problem by storing the result of frequent remote calls on the server, so that if the same data is needed again, it can accesss it from the cache, instead of making a remote call.
+
+# Types of caching
+There are 2 types of caching that `JITServer` does: global (in persistent memory) and per-compilation (on the heap). Both types of caching are done on per-client basis, that is, if multiple clients are connected to the same server, they will not share caches, as that would make things very complicated.
+Global caching is done for things that will not change (or very unlikely to change) over the lifetime of a client JVM, e.g. GC mode, IProfiler data for compiled methods.
+Local caching is done for things that are not going to change during the current compilation, but might change in-between compilations or are just unique for each compilation, e.g. resolved methods, IProfiler data for interpreted methods.
+Whenever possible, caching should be done globally, because hit rates will be higher, but one should be careful and make sure that the client data will not actually change.
+
+# Important caches
+## `ClientSessionData`
+Stores all of the globally cached data, for a detailed description read [this](https://github.com/eclipse/openj9/blob/jitaas/doc/compiler/jitaas/ClientSession.md).
+## Caches in `CompilationInfoPerThreadRemote`
+All of the per-compilation caches are stored in `CompilationInfoPerThreadRemote`. Since most caches have very similar structure, i.e. hash map, we added templated methods for working with these caches. The methods are `initializePerCompilationCache`, `cacheToPerCompilationMap`, `getCachedValueFromPerCompilationMap`, `clearPerCompilationCache`. If a new cache is added, it should use these methods. At the end of compilation all local caches need to be reset by adding a call to `clearPerCompilationCache` to `TR::CompilationInfoPerThreadRemote::clearPerCompilationCaches`.
```

Comment:
Is"`All of the per-compilation caches are stored in CompilationInfoPerThreadRemote` " an accurate statement? since down below you also mentioned "`Some per-compilation caches are stored inside resolved methods`".

Reply:
Yeah, true. I thought they were all stored in `CompilationInfoPerThreadRemote` but then I looked at the code and saw that some of them are stored in resolved methods.